### PR TITLE
Feature/1090 native drag and drop for file and folder

### LINKED ISSUE #1090: [FEAT]: Native drap and drop for file and folder with complete filepath
**Is your feature request related to a problem? Please describe.**
Currently, on Wails v1.16.8 as for windows development it depends on IE11 (as Windows development use the mshtml library which is only compatible with IE11), and On IE11, the drag and drop is not supported for folder. Whereas with Chrome, we can have file or folder drag and drop functionality.

**Describe the solution you'd like**
Feature that can be suitable would be native drag and drop with absolute path of dropped file/folder and if the dependency to IE11 can be removed.

**Describe alternatives you've considered**
Currently, We have tried the following approaches but not complete solution was found:
1. Use [ngx-file-drop](https://www.npmjs.com/package/ngx-file-drop) library for drag and drop, it worked fine for chrome but for IE11, only file drop was working fine and not folder. But was not providing absolute path of neither file not folder..
2. HTML Drag and drop: Failed for both file and folder.
3. Drag and drop on Input Type=File: Failed for both file and folder.
4. Created our own directive even it failed for both file and folder.

**Additional context**
Screenshot from example Feature Request

### PR DESCRIPTION
**Pull request by lyimmi**
# Description

This PR implement Drag & Drop for files and folders on Windows and Linux.

On both systems the whole window is a drop target and the developer can either use the JS runtime method to handle the drops as described below, or implement the handling them self.

I tried to test all of the cases I can think of but and extensive test by multiple people would be a good idea. Or maybe somehow make it experimental for a while. I don't really know how it wold be possible with the JavaScript code/files.

Fixes #1090

### Windows

The windows implementation is based on [ayatkyo](https://github.com/ayatkyo)'s work in https://github.com/wailsapp/wails/pull/2774, with minor changes to align the API with the linux imlementation.

The drop event sends the coordinates and the files to the go side to resolve the file paths and after that a `wails:file-drop` event is fired to notify the listeners in GO and Javascript.

### Linux

On Linux I have overridden the WebKitGTK signals for drag & drop and the C code sends the coordinates and the file paths via `processMessage()` to GO and than the same `wails:file-drop` event is fired.

### macOS

No solution for macOS.

#### GO  

```golang
OnFileDrop(ctx context.Context, callback func(x, y int, paths []string))
```

This can be used to handle the `wails:file-drop` event with a callback that handles the casting and checking the event's `interface{}` data.

```golang
OnFileDropOff(ctx context.Context) 
```

Remove the listener.

#### Javascript  

```javascript
OnFileDrop(callback: (x: number, y: number, paths: string[]) => void, useDropTarget: boolean) :void
```
Calls the callback function with the coordinates inside the window where the drag was released and a slice of absolute file paths.

When the `useDropTarget` is `true` in addition to calling the callback when the drop happens, it registers event listeners on
the window that are listening for the drag coordinates and checks if the mouse is over an element that has the
CSSDropProperty style. If the element has the required property
it adds the `wails-drop-target-active` class to the element's class list and removes it when the mouse moves off of it.

```javascript
OnFileDropOff(): void
```

This method removes all registered listeners and handlers for drag and drop events.

### Without runtime methods

You can simply listen to the `wails:file-drop` and handle it as you like.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
  
- [x] Windows 10
- [x] Windows 11
- [x] Linux Ubuntu 20.04 - X
- [x] Linux Ubuntu 23.10 - X
- [x] Linux Ubuntu 23.10 - Wayland

I tested it on multiple machines. On all systems I tested dropping in files and folders a combination of them slowly and fast, in-and-out. Sorry cannot describe my "method" better.

  
## Test Configuration

Windows:

```shell
# Wails
Version  | v2.7.1
Revision | e49e8894edd298dcffcfa4523f7a5ed469d75c33
Modified | true

# System
┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
| OS           | Windows 10 Pro                                                                    |
| Version      | 2009 (Build: 22631)                                                               |
| ID           | 23H2                                                                              |
| Go Version   | go1.21.5                                                                          |
| Platform     | windows                                                                           |
| Architecture | amd64                                                                             |
| CPU          | AMD Ryzen 5 6600H with Radeon Graphics                                            |
| GPU 1        | AMD Radeon(TM) Graphics (Advanced Micro Devices, Inc.) - Driver: 31.0.12046.13002 |
| GPU 2        | NVIDIA GeForce RTX 3060 Laptop GPU (NVIDIA) - Driver: 31.0.15.4633                |
| Memory       | 16GB                                                                              |
└──────────────────────────────────────────────────────────────────────────────────────────────────┘

# Dependencies
┌────────────────────────────────────────────────────────┐
| Dependency | Package Name | Status    | Version        |
| WebView2   | N/A          | Installed | 120.0.2210.133 |
| Nodejs     | N/A          | Installed | 18.18.0        |
| npm        | N/A          | Installed | 10.1.0         |
| *upx       | N/A          | Installed | upx 4.2.2      |
| *nsis      | N/A          | Available |                |
└─────────────── * - Optional Dependency ────────────────┘

# Diagnosis
Optional package(s) installation details:
  - nsis : More info at https://wails.io/docs/guides/windows-installer/

 SUCCESS  Your system is ready for Wails development!
```

Linux:

```shell
# Wails
Version         | v2.7.1                                  
Revision        | 66f82e9cdb894486f6df1de4cbf55c465c8912bf
Modified        | true                                    
Package Manager | apt                                     

# System
┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
| OS           | Ubuntu                                                                                           |
| Version      | 23.10                                                                                            |
| ID           | ubuntu                                                                                           |
| Go Version   | go1.21.6                                                                                         |
| Platform     | linux                                                                                            |
| Architecture | amd64                                                                                            |
| CPU          | AMD Ryzen 7 3700X 8-Core Processor                                                               |
| GPU          | Navi 23 [Radeon RX 6600/6600 XT/6600M] (Advanced Micro Devices, Inc. [AMD/ATI]) - Driver: amdgpu |
| Memory       | 32GB                                                                                             |
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

# Dependencies
┌──────────────────────────────────────────────────────────────────────────┐
| Dependency | Package Name          | Status    | Version                 |
| *docker    | docker.io             | Installed | 24.0.7                  |
| gcc        | build-essential       | Installed | 12.10ubuntu1            |
| libgtk-3   | libgtk-3-dev          | Installed | 3.24.38-5ubuntu1        |
| libwebkit  | libwebkit2gtk-4.0-dev | Installed | 2.42.4-0ubuntu0.23.10.1 |
| npm        | npm                   | Installed | 9.8.0                   |
| *nsis      | nsis                  | Installed | v3.09-1                 |
| pkg-config | pkg-config            | Installed | 1.8.1-2                 |
└──────────────────────── * - Optional Dependency ─────────────────────────┘

# Diagnosis
 SUCCESS  Your system is ready for Wails development!

```

# Checklist:

- [x] I have updated `website/src/pages/changelog.mdx` with details of this PR
- [x] My code follows the general coding style of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Summary by CodeRabbit

- **New Features**
  - Introduced drag and drop support for files and folders on Windows and Linux platforms.
  - Implemented new JavaScript commands for setting CSS properties related to drag and drop.

- **Documentation**
  - Updated options documentation to include new `DragAndDrop` configuration fields.
  - Added new runtime documentation for handling drag and drop operations.

- **Enhancements**
  - Improved file drag and drop event handling with new callback registration functions.
  - Enhanced desktop runtime environment with drag and drop event management.

- **Bug Fixes**
  - Removed quarantine attribute on macOS binaries to address execution issues.
  - Added documentation for a common GStreamer error on Linux systems.
  - Fixed dock icon right-click exit handling for improved user experience.

### CODE REVIEW

File: website/src/pages/changelog.mdx [hunk 1]
```diff
@@ -21,6 +21,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Remove quarantine attribute on macOS binaries. Changed by @leaanthony in [PR](https://github.com/wailsapp/wails/pull/3118)
 - Added documentation for a common GStreamer error on Linux systems. Changed by [@mkwsnyder](https://github.com/mkwsnyder) in [PR](https://github.com/wailsapp/wails/pull/3134)
 - Add  dock icon right-click exit handling by @almas1992 in [PR](https://github.com/wailsapp/wails/pull/3157)
+- Added Drag & Drop (files or folders) support for Windows and Linux. Added by [@lyimmi](https://github.com/lyimmi) in [PR](?). Based on the work of [@ayatkyo](https://github.com/ayatkyo) for Windows in [PR](https://github.com/wailsapp/wails/pull/2774).
```
[the remark this answers is not in the document]
Comment (lyimmi) [reply website/src/pages/changelog.mdx]: Resolved after PR was sent.

File: v2/internal/frontend/desktop/linux/window.c [hunk 2]
```diff
@@ -198,7 +199,7 @@ static gboolean setTitle(gpointer data)
 
 void SetTitle(GtkWindow *window, char *title)
 {
-    SetTitleArgs *args = malloc(sizeof(SetTitleArgs));
+    SetTitleArgs *args = (char *)malloc(sizeof(SetTitleArgs));
```
[the remark this answers is not in the document]
Comment (lyimmi) [reply v2/internal/frontend/desktop/linux/window.c]: I don't remember doing that... It was a mistake.
Checked the git history and reverted that change in bbef47d to the previous state.

**Comment by lyimmi** [comment]
@coderabbitai review

**Comment by risico** [comment]
Hi @lyimmi! I was wondering if I could help you add this for MacOS.  Reading from your Issue thread:

```
On mac I think the implementation would be the same as on Linux. But I'm not versed in the magic of macOs develpement.
```

But in this PR we've:

```
macOS
No solution for macOS.
```

I assume a solution would be possible but you haven't looked into it, or you actually meant there's no solution for MacOS?

**Comment by lyimmi** [comment]
@risico I meant I have no solution for macOS. That would be nice if you could help with the macOS implementation!

**Comment by pavelbinar** [comment]
**Hi @lyimmi, this is fantastics!** Thank you for pushing this through.

@leaanthony If this is merged do you plan to put it into v2 or v3?

I am asking because my team is considering adding support for Mac (if @lyimmi can guide us a bit), but we would love to see it in production sooner then later.

CC: @jakubpeleska

**Comment by leaanthony** [comment]
@pavelbinar - I'm ok with this going into v2. There is a solution for v3 and I wanted to diff to see what the differences were. I'd prefer there to be minimum api/concept changes between v2 and v3 WRT this feature. The v3 example shows the API that's currently implemented: https://github.com/wailsapp/wails/blob/8463c01123635050bce13f738e38dd9891438558/v3/examples/drag-n-drop/main.go 

The implementation for mac is here: https://github.com/wailsapp/wails/blob/dc8cbcf41047e5690e688922cd3c058f9428626b/v3/pkg/application/webview_window_darwin_drag.m

**Comment by leaanthony** [comment]
@lyimmi - did this need any updates to `go-webview2`?

**Comment by lyimmi** [comment]
@leaanthony - no I just checked out the master and everything I needed was already there, didn't touched any dependecies.

**Comment by lyimmi** [comment]
I'm not too familiar how github works regarding pushing changes to a PR from a fork. Should @pavelbinar push changes to my fork? 

Should I set this PR to draft until the mac side is done?

File: v2/pkg/runtime/draganddrop.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		if !ok {
+			LogError(ctx, fmt.Sprintf("invalid y coordinate in drag and drop: %v", optionalData[1]))
+		}
+		paths, ok := optionalData[2].([]string)
+		if !ok {
+			LogError(ctx, fmt.Sprintf("invalid path data in drag and drop: %v", optionalData[2]))
+		}
+		callback(x, y, paths)
+	})
+}
```
[the remark this answers is not in the document]
Comment (lyimmi) [reply v2/pkg/runtime/draganddrop.go]: changed in commit: 209d74d

File: v2/pkg/runtime/draganddrop.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+)
+
+// OnFileDrop returns a slice of file path strings when a drop is finished.
+func OnFileDrop(ctx context.Context, callback func(x, y int, paths []string)) {
+	EventsOn(ctx, "wails:file-drop", func(optionalData ...interface{}) {
```
[the remark this answers is not in the document]
Comment (lyimmi) [reply v2/pkg/runtime/draganddrop.go:13]: OnFileDrop function handles casting and type checking before calling the callback function.

File: v2/pkg/runtime/draganddrop.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+		if callback == nil {
+			return
+		}
+		if len(optionalData) != 3 {
+			callback(0, 0, nil)
```
[the remark this answers is not in the document]
Comment (lyimmi) [reply v2/pkg/runtime/draganddrop.go:15]: I think it is a sane default behavior in this case.

File: v2/pkg/runtime/draganddrop.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		if !ok {
+			LogError(ctx, fmt.Sprintf("invalid x coordinate in drag and drop: %v", optionalData[0]))
+		}
+		y, ok := optionalData[1].(int)
+		if !ok {
+			LogError(ctx, fmt.Sprintf("invalid y coordinate in drag and drop: %v", optionalData[1]))
+		}
+		paths, ok := optionalData[2].([]string)
+		if !ok {
+			LogError(ctx, fmt.Sprintf("invalid path data in drag and drop: %v", optionalData[2]))
```
[the remark this answers is not in the document]
Comment (lyimmi) [reply v2/pkg/runtime/draganddrop.go:27]: The logging follows the style of code for event handlers.

File: v2/pkg/runtime/draganddrop.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lyimmi) [reply v2/pkg/runtime/draganddrop.go]: I think it's a valid behavior to send zero values when the coordinates cannot be resolved or when there is no file path.

**Comment by leaanthony** [comment]
Having fun @lyimmi ? 😉

**Comment by lyimmi** [comment]
@leaanthony Yeah. It's pretty good actually. But I'm done now... Sorry. 😁

**Comment by pavelbinar** [comment]
@leaanthony - Thank you for the clarification.

@lyimmi - I appreciate your willingness to help. We will get in touch within a few weeks (may we connect via email?). I will reach out to you later, once we are more familiar with the code and have some questions. Currently, I am exploring the options and preparing the work.

Regarding the PR:
Ideally, @leaanthony would merge your PR into a separate feature branch (instead of master), and later on, we will submit a separate PR based on this feature branch. This way, once all platforms are accounted for, @leaanthony can merge the complete feature branch into master and celebrate with a 🍺

**Comment by leaanthony** [comment]
I've created `feature/1090_native_drag_and_drop_for_file_and_folder` for this :thu

**Comment by leaanthony** [comment]
I've created `feature/1090_native_drag_and_drop_for_file_and_folder` for this 👍

**Comment by pavelbinar** [comment]
@lyimmi 🎉 
Well done

@leaanthony perfect 👌🏻

**Pull request merged**